File: day12_part2_2018.php

<?php

$file = fopen("input.txt", "r") or die("Unable to open file!");
$initialState = "";
$rules = [];

while (!feof($file)) {
    $line = trim(fgets($file));
    if (strpos($line, "initial state") !== false) {
        $initialState = explode(": ", $line)[1];
    } elseif (strpos($line, "=>") !== false) {
        $parts = explode(" => ", $line);
        $rules[$parts[0]] = $parts[1][0];
    }
}

$state = [];
for ($i = 0; $i < strlen($initialState); $i++) {
    if ($initialState[$i] === '#') {
        $state[$i] = '#';
    }
}

$previousPattern = "";
$previousSum = 0;
$offset = 0;
for ($generation = 0; $generation < 50000000000; $generation++) {
    $newState = [];
    $minPot = min(array_keys($state)) - 2;
    $maxPot = max(array_keys($state)) + 2;
    for ($i = $minPot; $i <= $maxPot; $i++) {
        $pattern = "";
        for ($j = $i - 2; $j <= $i + 2; $j++) {
            $pattern .= isset($state[$j]) && $state[$j] === '#' ? '#' : '.';
        }
        if (isset($rules[$pattern]) && $rules[$pattern] === '#') {
            $newState[$i] = '#';
        }
    }
    $state = $newState;

    list($currentPattern, $currentSum) = statePattern($state);
    if ($currentPattern === $previousPattern) {
        $offset = $currentSum - $previousSum;
        $remainingGenerations = 50000000000 - $generation - 1;
        $finalSum = $currentSum + $offset * $remainingGenerations;
        echo $finalSum . PHP_EOL;
        break;
    }
    $previousPattern = $currentPattern;
    $previousSum = $currentSum;
}

function statePattern($state) {
    $minPot = min(array_keys($state));
    $maxPot = max(array_keys($state));
    $pattern = "";
    $sum = 0;
    for ($i = $minPot; $i <= $maxPot; $i++) {
        if (isset($state[$i]) && $state[$i] === '#') {
            $pattern .= '#';
            $sum += $i;
        } else {
            $pattern .= '.';
        }
    }
    return [$pattern, $sum];
}

fclose($file);
?>

File: matlab/day12_part2_2018.m

function main()
    fid = fopen('input.txt', 'r');
    
    line = fgetl(fid);
    initial_state_str = sscanf(line, 'initial state: %s');
    
    fgetl(fid);
    
    rules = containers.Map('KeyType', 'char', 'ValueType', 'logical');
    while ~feof(fid)
        line = fgetl(fid);
        if isempty(line)
            continue;
        end
        parts = sscanf(line, '%s => %c');
        if parts(end) == '#'
            rules(parts(1:5)) = true;
        end
    end
    fclose(fid);

    plant_indices = find(initial_state_str == '#') - 1;

    total_generations = 50000000000;
    previous_pattern = [];
    previous_sum = 0;

    for gen = 1:200
        min_idx = min(plant_indices);
        max_idx = max(plant_indices);
        
        new_plant_indices = [];
        
        for i = (min_idx - 2):(max_idx + 2)
            pattern_indices = (i-2):(i+2);
            has_plant = ismember(pattern_indices, plant_indices);
            pattern_str = char(has_plant * ('#' - '.') + '.');
            
            if isKey(rules, pattern_str)
                new_plant_indices(end+1) = i;
            end
        end
        
        plant_indices = new_plant_indices;
        
        current_sum = sum(plant_indices);
        current_pattern = plant_indices - min(plant_indices);
        
        if isequal(current_pattern, previous_pattern)
            offset = current_sum - previous_sum;
            remaining_generations = total_generations - gen;
            final_sum = current_sum + offset * remaining_generations;
            
            fprintf('%.0f\n', final_sum);
            return;
        end
        
        previous_pattern = current_pattern;
        previous_sum = current_sum;
    end
end
